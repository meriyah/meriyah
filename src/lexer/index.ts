export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
