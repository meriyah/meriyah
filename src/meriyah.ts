import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
