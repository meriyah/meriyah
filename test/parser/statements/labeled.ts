import * as t from 'node:assert/strict';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail, pass } from '../../test-utils';

describe('Statements - Labeled', () => {
  for (const arg of [
    'break',
    'case',
    'catch',
    'class',
    'const',
    'continue',
    'debugger',
    'default',
    'delete',
    'do',
    'else',
    'export',
    'extends',
    'finally',
    'for',
    'function',
    'if',
    'import',
    'in',
    'instanceof',
    'new',
    'return',
    'super',
    'switch',
    'this',
    'throw',
    'try',
    'typeof',
    'var',
    'void',
    'while',
    'with',
    'null',
    'true',
    'false',
    'enum',
  ]) {
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg} : x`);
      });
    });
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg} : x`, { lexical: true });
      });
    });
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg} : x`, { webcompat: true });
      });
    });
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg} : x`, { impliedStrict: true });
      });
    });
  }

  fail('Statements - Labeled (fail)', [
    'label: class C {};',
    'label: let x;',
    'a: async function* a(){}',
    'if (false) label1: label2: function test262() {} else ;',
    'label: function* g() {}',
    'label: const x = null;',
    { code: 'label: function g() {}', options: { impliedStrict: true } },
    'label: let x;',
    { code: 'await: 1;', options: { sourceType: 'module' } },
    'bar: function* x() {}',
    { code: 'await: 1;', options: { sourceType: 'module' } },
    { code: 'yield: 1;', options: { impliedStrict: true } },
    'foo:for;',
    'super: while(true) { break super; }"',
    'function test_func() { super: while(true) { break super; }}"',
    '() => {super: while(true) { break super; }}"',
    'do { test262: { continue test262; } } while (false)',
    '"use strict"; super: while(true) { break super; }',
    '"use strict"; package: while(true) { break package; }',
    'false: ;',
    'true: ;',
    '(async function*() { yield: 1; });',
    'function *gen() { yield: ;}',
    { code: 'function *gen() { yield: ;}', options: { impliedStrict: true } },
    'var obj = { *method() { yield: ; } };',
    { code: 'var obj = { *method() { yield: ; } };', options: { impliedStrict: true } },
    'foo: function f() {}',
    String.raw`async () => { \u{61}wait: x }`,
    String.raw`async () => { aw\u{61}it: x }`,
    { code: String.raw`async () => { \u{61}wait: x }`, options: { sourceType: 'module' } },
    String.raw`async () => { aw\u{61}it: x }`,
    { code: 'function *f(){ await: x; }', options: { sourceType: 'module' } },
    { code: 'await: x', options: { sourceType: 'module' } },
    { code: 'await: 1;', options: { sourceType: 'module' } },
    'false: x',
    { code: 'implements: x', options: { impliedStrict: true } },
    { code: 'package: x', options: { impliedStrict: true } },
    { code: 'let: x', options: { impliedStrict: true } },
    { code: 'yield: x', options: { impliedStrict: true } },
    { code: 'function *f(){ yield: x; }', options: { impliedStrict: true } },
    { code: 'yield: { function *f(){ break await; } }', options: { impliedStrict: true } },
    'bar: foo: ding: foo: x',
    'foo: bar: foo: x',
    'a: { a: x }',
    'yield: { function *f(){ break await; } }',
    'yield: { function *f(){ break await; } }',
  ]);

  pass('Statements - Labeled (pass)', [
    { code: 'function *f(){ await: x; }', options: { webcompat: true, ranges: true } },
    { code: 'await: while (await) { continue await; }', options: { webcompat: true, ranges: true } },
    { code: 'async: while (async) { continue async; }', options: { webcompat: true } },
    { code: 'let, let, let, let', options: { webcompat: true, ranges: true } },
    { code: 'let: foo', options: { webcompat: true, ranges: true } },
    { code: 'foo: function bar() {}', options: { webcompat: true, ranges: true } },
    'yield: await',
    'a:package',
    '__proto__: test',
    { code: 'a:{break a;}', options: { ranges: true } },
    'async: await',
    { code: 'start: while (true) break start', options: { ranges: true } },
    'if (false) {\n L: let\nx = 1; \n }',
    { code: 'foo: bar: function f(){}', options: { webcompat: true } },
    { code: 'foo: bar: third: function f(){}', options: { webcompat: true } },
    'L: let\nx',
    { code: '__proto__: while (true) { break __proto__; }', options: { ranges: true } },
    'a:{break a;}',
  ]);
});
