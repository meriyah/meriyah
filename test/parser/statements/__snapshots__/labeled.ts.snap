// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Statements - Labeled > Statements - Labeled (fail) > "use strict"; package: while(true) { break package; } 1`] = `
"SyntaxError [1:14-1:21]: Unexpected token: 'package'
> 1 | "use strict"; package: while(true) { break package; }
    |               ^^^^^^^ Unexpected token: 'package'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > "use strict"; super: while(true) { break super; } 1`] = `
"SyntaxError [1:19-1:20]: Unexpected token: 'super'
> 1 | "use strict"; super: while(true) { break super; }
    |                    ^ Unexpected token: 'super'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > () => {super: while(true) { break super; }}" 1`] = `
"SyntaxError [1:12-1:13]: Unexpected token: 'super'
> 1 | () => {super: while(true) { break super; }}"
    |             ^ Unexpected token: 'super'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > (async function*() { yield: 1; }); 1`] = `
"SyntaxError [1:26-1:27]: 'yield' may not be used as an identifier in this context
> 1 | (async function*() { yield: 1; });
    |                           ^ 'yield' may not be used as an identifier in this context"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > a: { a: x } 1`] = `
"SyntaxError [1:6-1:7]: Label 'a' has already been declared
> 1 | a: { a: x }
    |       ^ Label 'a' has already been declared"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > a: async function* a(){} 1`] = `
"SyntaxError [1:9-1:17]: Async functions can only be declared at the top level or inside a block
> 1 | a: async function* a(){}
    |          ^^^^^^^^ Async functions can only be declared at the top level or inside a block"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > async () => { \\u{61}wait: x } 1`] = `
"SyntaxError [1:14-1:24]: Unexpected token: 'escaped keyword'
> 1 | async () => { \\u{61}wait: x }
    |               ^^^^^^^^^^ Unexpected token: 'escaped keyword'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > async () => { \\u{61}wait: x } 2`] = `
"SyntaxError [1:14-1:24]: Unexpected token: 'escaped keyword'
> 1 | async () => { \\u{61}wait: x }
    |               ^^^^^^^^^^ Unexpected token: 'escaped keyword'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > async () => { aw\\u{61}it: x } 1`] = `
"SyntaxError [1:14-1:24]: Unexpected token: 'escaped keyword'
> 1 | async () => { aw\\u{61}it: x }
    |               ^^^^^^^^^^ Unexpected token: 'escaped keyword'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > async () => { aw\\u{61}it: x } 2`] = `
"SyntaxError [1:14-1:24]: Unexpected token: 'escaped keyword'
> 1 | async () => { aw\\u{61}it: x }
    |               ^^^^^^^^^^ Unexpected token: 'escaped keyword'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > await: 1; 1`] = `
"SyntaxError [1:0-1:5]: Can not use \`await\` as identifier in module or async func
> 1 | await: 1;
    | ^^^^^ Can not use \`await\` as identifier in module or async func"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > await: 1; 2`] = `
"SyntaxError [1:0-1:5]: Can not use \`await\` as identifier in module or async func
> 1 | await: 1;
    | ^^^^^ Can not use \`await\` as identifier in module or async func"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > await: 1; 3`] = `
"SyntaxError [1:0-1:5]: Can not use \`await\` as identifier in module or async func
> 1 | await: 1;
    | ^^^^^ Can not use \`await\` as identifier in module or async func"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > await: x 1`] = `
"SyntaxError [1:0-1:5]: Can not use \`await\` as identifier in module or async func
> 1 | await: x
    | ^^^^^ Can not use \`await\` as identifier in module or async func"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > bar: foo: ding: foo: x 1`] = `
"SyntaxError [1:19-1:20]: Label 'foo' has already been declared
> 1 | bar: foo: ding: foo: x
    |                    ^ Label 'foo' has already been declared"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > bar: function* x() {} 1`] = `
"SyntaxError [1:5-1:13]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
> 1 | bar: function* x() {}
    |      ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > do { test262: { continue test262; } } while (false) 1`] = `
"SyntaxError [1:32-1:33]: continue statement must be nested within an iteration statement
> 1 | do { test262: { continue test262; } } while (false)
    |                                 ^ continue statement must be nested within an iteration statement"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > false: ; 1`] = `
"SyntaxError [1:5-1:6]: Invalid keyword
> 1 | false: ;
    |      ^ Invalid keyword"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > false: x 1`] = `
"SyntaxError [1:5-1:6]: Invalid keyword
> 1 | false: x
    |      ^ Invalid keyword"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > foo: bar: foo: x 1`] = `
"SyntaxError [1:13-1:14]: Label 'foo' has already been declared
> 1 | foo: bar: foo: x
    |              ^ Label 'foo' has already been declared"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > foo: function f() {} 1`] = `
"SyntaxError [1:5-1:13]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
> 1 | foo: function f() {}
    |      ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > foo:for; 1`] = `
"SyntaxError [1:7-1:8]: Expected '('
> 1 | foo:for;
    |        ^ Expected '('"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > function *f(){ await: x; } 1`] = `
"SyntaxError [1:15-1:20]: Can not use \`await\` as identifier in module or async func
> 1 | function *f(){ await: x; }
    |                ^^^^^ Can not use \`await\` as identifier in module or async func"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > function *f(){ yield: x; } 1`] = `
"SyntaxError [1:20-1:21]: Unexpected strict mode reserved word
> 1 | function *f(){ yield: x; }
    |                     ^ Unexpected strict mode reserved word"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > function *gen() { yield: ;} 1`] = `
"SyntaxError [1:23-1:24]: 'yield' may not be used as an identifier in this context
> 1 | function *gen() { yield: ;}
    |                        ^ 'yield' may not be used as an identifier in this context"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > function *gen() { yield: ;} 2`] = `
"SyntaxError [1:23-1:24]: Unexpected strict mode reserved word
> 1 | function *gen() { yield: ;}
    |                        ^ Unexpected strict mode reserved word"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > function test_func() { super: while(true) { break super; }}" 1`] = `
"SyntaxError [1:28-1:29]: Unexpected token: 'super'
> 1 | function test_func() { super: while(true) { break super; }}"
    |                             ^ Unexpected token: 'super'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > if (false) label1: label2: function test262() {} else ; 1`] = `
"SyntaxError [1:27-1:35]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
> 1 | if (false) label1: label2: function test262() {} else ;
    |                            ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > implements: x 1`] = `
"SyntaxError [1:0-1:10]: Unexpected token: 'implements'
> 1 | implements: x
    | ^^^^^^^^^^ Unexpected token: 'implements'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > label: class C {}; 1`] = `
"SyntaxError [1:7-1:12]: Class declaration can't appear in single-statement context
> 1 | label: class C {};
    |        ^^^^^ Class declaration can't appear in single-statement context"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > label: const x = null; 1`] = `
"SyntaxError [1:7-1:12]: Unexpected token: 'const'
> 1 | label: const x = null;
    |        ^^^^^ Unexpected token: 'const'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > label: function g() {} 1`] = `
"SyntaxError [1:7-1:15]: In strict mode code, functions can only be declared at top level or inside a block
> 1 | label: function g() {}
    |        ^^^^^^^^ In strict mode code, functions can only be declared at top level or inside a block"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > label: function* g() {} 1`] = `
"SyntaxError [1:7-1:15]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
> 1 | label: function* g() {}
    |        ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > label: let x; 1`] = `
"SyntaxError [1:11-1:12]: Unexpected token: 'identifier'
> 1 | label: let x;
    |            ^ Unexpected token: 'identifier'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > label: let x; 2`] = `
"SyntaxError [1:11-1:12]: Unexpected token: 'identifier'
> 1 | label: let x;
    |            ^ Unexpected token: 'identifier'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > let: x 1`] = `
"SyntaxError [1:3-1:4]: \`let\` declaration not allowed here and \`let\` cannot be a regular var name in strict mode
> 1 | let: x
    |    ^ \`let\` declaration not allowed here and \`let\` cannot be a regular var name in strict mode"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > package: x 1`] = `
"SyntaxError [1:0-1:7]: Unexpected token: 'package'
> 1 | package: x
    | ^^^^^^^ Unexpected token: 'package'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > super: while(true) { break super; }" 1`] = `
"SyntaxError [1:5-1:6]: Unexpected token: 'super'
> 1 | super: while(true) { break super; }"
    |      ^ Unexpected token: 'super'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > true: ; 1`] = `
"SyntaxError [1:4-1:5]: Invalid keyword
> 1 | true: ;
    |     ^ Invalid keyword"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > var obj = { *method() { yield: ; } }; 1`] = `
"SyntaxError [1:29-1:30]: 'yield' may not be used as an identifier in this context
> 1 | var obj = { *method() { yield: ; } };
    |                              ^ 'yield' may not be used as an identifier in this context"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > var obj = { *method() { yield: ; } }; 2`] = `
"SyntaxError [1:29-1:30]: Unexpected strict mode reserved word
> 1 | var obj = { *method() { yield: ; } };
    |                              ^ Unexpected strict mode reserved word"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > yield: { function *f(){ break await; } } 1`] = `
"SyntaxError [1:0-1:5]: 'yield' may not be used as an identifier in this context
> 1 | yield: { function *f(){ break await; } }
    | ^^^^^ 'yield' may not be used as an identifier in this context"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > yield: { function *f(){ break await; } } 2`] = `
"SyntaxError [1:35-1:36]: Undefined label 'await'
> 1 | yield: { function *f(){ break await; } }
    |                                    ^ Undefined label 'await'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > yield: { function *f(){ break await; } } 3`] = `
"SyntaxError [1:35-1:36]: Undefined label 'await'
> 1 | yield: { function *f(){ break await; } }
    |                                    ^ Undefined label 'await'"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > yield: 1; 1`] = `
"SyntaxError [1:0-1:5]: 'yield' may not be used as an identifier in this context
> 1 | yield: 1;
    | ^^^^^ 'yield' may not be used as an identifier in this context"
`;

exports[`Statements - Labeled > Statements - Labeled (fail) > yield: x 1`] = `
"SyntaxError [1:0-1:5]: 'yield' may not be used as an identifier in this context
> 1 | yield: x
    | ^^^^^ 'yield' may not be used as an identifier in this context"
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > __proto__: test 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "test",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "label": {
        "name": "__proto__",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > __proto__: while (true) { break __proto__; } 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "body": [
            {
              "end": 42,
              "label": {
                "end": 41,
                "name": "__proto__",
                "range": [
                  32,
                  41,
                ],
                "start": 32,
                "type": "Identifier",
              },
              "range": [
                26,
                42,
              ],
              "start": 26,
              "type": "BreakStatement",
            },
          ],
          "end": 44,
          "range": [
            24,
            44,
          ],
          "start": 24,
          "type": "BlockStatement",
        },
        "end": 44,
        "range": [
          11,
          44,
        ],
        "start": 11,
        "test": {
          "end": 22,
          "range": [
            18,
            22,
          ],
          "start": 18,
          "type": "Literal",
          "value": true,
        },
        "type": "WhileStatement",
      },
      "end": 44,
      "label": {
        "end": 9,
        "name": "__proto__",
        "range": [
          0,
          9,
        ],
        "start": 0,
        "type": "Identifier",
      },
      "range": [
        0,
        44,
      ],
      "start": 0,
      "type": "LabeledStatement",
    },
  ],
  "end": 44,
  "range": [
    0,
    44,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > L: let
x 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "let",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "label": {
        "name": "L",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
    {
      "expression": {
        "name": "x",
        "type": "Identifier",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > a:{break a;} 1`] = `
{
  "body": [
    {
      "body": {
        "body": [
          {
            "end": 11,
            "label": {
              "end": 10,
              "name": "a",
              "range": [
                9,
                10,
              ],
              "start": 9,
              "type": "Identifier",
            },
            "range": [
              3,
              11,
            ],
            "start": 3,
            "type": "BreakStatement",
          },
        ],
        "end": 12,
        "range": [
          2,
          12,
        ],
        "start": 2,
        "type": "BlockStatement",
      },
      "end": 12,
      "label": {
        "end": 1,
        "name": "a",
        "range": [
          0,
          1,
        ],
        "start": 0,
        "type": "Identifier",
      },
      "range": [
        0,
        12,
      ],
      "start": 0,
      "type": "LabeledStatement",
    },
  ],
  "end": 12,
  "range": [
    0,
    12,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > a:{break a;} 2`] = `
{
  "body": [
    {
      "body": {
        "body": [
          {
            "label": {
              "name": "a",
              "type": "Identifier",
            },
            "type": "BreakStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "label": {
        "name": "a",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > a:package 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "package",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "label": {
        "name": "a",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > async: await 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "await",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "label": {
        "name": "async",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > async: while (async) { continue async; } 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "body": [
            {
              "label": {
                "name": "async",
                "type": "Identifier",
              },
              "type": "ContinueStatement",
            },
          ],
          "type": "BlockStatement",
        },
        "test": {
          "name": "async",
          "type": "Identifier",
        },
        "type": "WhileStatement",
      },
      "label": {
        "name": "async",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > await: while (await) { continue await; } 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "body": [
            {
              "end": 38,
              "label": {
                "end": 37,
                "name": "await",
                "range": [
                  32,
                  37,
                ],
                "start": 32,
                "type": "Identifier",
              },
              "range": [
                23,
                38,
              ],
              "start": 23,
              "type": "ContinueStatement",
            },
          ],
          "end": 40,
          "range": [
            21,
            40,
          ],
          "start": 21,
          "type": "BlockStatement",
        },
        "end": 40,
        "range": [
          7,
          40,
        ],
        "start": 7,
        "test": {
          "end": 19,
          "name": "await",
          "range": [
            14,
            19,
          ],
          "start": 14,
          "type": "Identifier",
        },
        "type": "WhileStatement",
      },
      "end": 40,
      "label": {
        "end": 5,
        "name": "await",
        "range": [
          0,
          5,
        ],
        "start": 0,
        "type": "Identifier",
      },
      "range": [
        0,
        40,
      ],
      "start": 0,
      "type": "LabeledStatement",
    },
  ],
  "end": 40,
  "range": [
    0,
    40,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > foo: bar: function f(){} 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "async": false,
          "body": {
            "body": [],
            "type": "BlockStatement",
          },
          "generator": false,
          "id": {
            "name": "f",
            "type": "Identifier",
          },
          "params": [],
          "type": "FunctionDeclaration",
        },
        "label": {
          "name": "bar",
          "type": "Identifier",
        },
        "type": "LabeledStatement",
      },
      "label": {
        "name": "foo",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > foo: bar: third: function f(){} 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "body": {
            "async": false,
            "body": {
              "body": [],
              "type": "BlockStatement",
            },
            "generator": false,
            "id": {
              "name": "f",
              "type": "Identifier",
            },
            "params": [],
            "type": "FunctionDeclaration",
          },
          "label": {
            "name": "third",
            "type": "Identifier",
          },
          "type": "LabeledStatement",
        },
        "label": {
          "name": "bar",
          "type": "Identifier",
        },
        "type": "LabeledStatement",
      },
      "label": {
        "name": "foo",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > foo: function bar() {} 1`] = `
{
  "body": [
    {
      "body": {
        "async": false,
        "body": {
          "body": [],
          "end": 22,
          "range": [
            20,
            22,
          ],
          "start": 20,
          "type": "BlockStatement",
        },
        "end": 22,
        "generator": false,
        "id": {
          "end": 17,
          "name": "bar",
          "range": [
            14,
            17,
          ],
          "start": 14,
          "type": "Identifier",
        },
        "params": [],
        "range": [
          5,
          22,
        ],
        "start": 5,
        "type": "FunctionDeclaration",
      },
      "end": 22,
      "label": {
        "end": 3,
        "name": "foo",
        "range": [
          0,
          3,
        ],
        "start": 0,
        "type": "Identifier",
      },
      "range": [
        0,
        22,
      ],
      "start": 0,
      "type": "LabeledStatement",
    },
  ],
  "end": 22,
  "range": [
    0,
    22,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > function *f(){ await: x; } 1`] = `
{
  "body": [
    {
      "async": false,
      "body": {
        "body": [
          {
            "body": {
              "end": 24,
              "expression": {
                "end": 23,
                "name": "x",
                "range": [
                  22,
                  23,
                ],
                "start": 22,
                "type": "Identifier",
              },
              "range": [
                22,
                24,
              ],
              "start": 22,
              "type": "ExpressionStatement",
            },
            "end": 24,
            "label": {
              "end": 20,
              "name": "await",
              "range": [
                15,
                20,
              ],
              "start": 15,
              "type": "Identifier",
            },
            "range": [
              15,
              24,
            ],
            "start": 15,
            "type": "LabeledStatement",
          },
        ],
        "end": 26,
        "range": [
          13,
          26,
        ],
        "start": 13,
        "type": "BlockStatement",
      },
      "end": 26,
      "generator": true,
      "id": {
        "end": 11,
        "name": "f",
        "range": [
          10,
          11,
        ],
        "start": 10,
        "type": "Identifier",
      },
      "params": [],
      "range": [
        0,
        26,
      ],
      "start": 0,
      "type": "FunctionDeclaration",
    },
  ],
  "end": 26,
  "range": [
    0,
    26,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > if (false) {
 L: let
x = 1; 
 } 1`] = `
{
  "body": [
    {
      "alternate": null,
      "consequent": {
        "body": [
          {
            "body": {
              "expression": {
                "name": "let",
                "type": "Identifier",
              },
              "type": "ExpressionStatement",
            },
            "label": {
              "name": "L",
              "type": "Identifier",
            },
            "type": "LabeledStatement",
          },
          {
            "expression": {
              "left": {
                "name": "x",
                "type": "Identifier",
              },
              "operator": "=",
              "right": {
                "type": "Literal",
                "value": 1,
              },
              "type": "AssignmentExpression",
            },
            "type": "ExpressionStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "test": {
        "type": "Literal",
        "value": false,
      },
      "type": "IfStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > let, let, let, let 1`] = `
{
  "body": [
    {
      "end": 18,
      "expression": {
        "end": 18,
        "expressions": [
          {
            "end": 3,
            "name": "let",
            "range": [
              0,
              3,
            ],
            "start": 0,
            "type": "Identifier",
          },
          {
            "end": 8,
            "name": "let",
            "range": [
              5,
              8,
            ],
            "start": 5,
            "type": "Identifier",
          },
          {
            "end": 13,
            "name": "let",
            "range": [
              10,
              13,
            ],
            "start": 10,
            "type": "Identifier",
          },
          {
            "end": 18,
            "name": "let",
            "range": [
              15,
              18,
            ],
            "start": 15,
            "type": "Identifier",
          },
        ],
        "range": [
          0,
          18,
        ],
        "start": 0,
        "type": "SequenceExpression",
      },
      "range": [
        0,
        18,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 18,
  "range": [
    0,
    18,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > let: foo 1`] = `
{
  "body": [
    {
      "body": {
        "end": 8,
        "expression": {
          "end": 8,
          "name": "foo",
          "range": [
            5,
            8,
          ],
          "start": 5,
          "type": "Identifier",
        },
        "range": [
          5,
          8,
        ],
        "start": 5,
        "type": "ExpressionStatement",
      },
      "end": 8,
      "label": {
        "end": 3,
        "name": "let",
        "range": [
          0,
          3,
        ],
        "start": 0,
        "type": "Identifier",
      },
      "range": [
        0,
        8,
      ],
      "start": 0,
      "type": "LabeledStatement",
    },
  ],
  "end": 8,
  "range": [
    0,
    8,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > start: while (true) break start 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "end": 31,
          "label": {
            "end": 31,
            "name": "start",
            "range": [
              26,
              31,
            ],
            "start": 26,
            "type": "Identifier",
          },
          "range": [
            20,
            31,
          ],
          "start": 20,
          "type": "BreakStatement",
        },
        "end": 31,
        "range": [
          7,
          31,
        ],
        "start": 7,
        "test": {
          "end": 18,
          "range": [
            14,
            18,
          ],
          "start": 14,
          "type": "Literal",
          "value": true,
        },
        "type": "WhileStatement",
      },
      "end": 31,
      "label": {
        "end": 5,
        "name": "start",
        "range": [
          0,
          5,
        ],
        "start": 0,
        "type": "Identifier",
      },
      "range": [
        0,
        31,
      ],
      "start": 0,
      "type": "LabeledStatement",
    },
  ],
  "end": 31,
  "range": [
    0,
    31,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Labeled > Statements - Labeled (pass) > yield: await 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "await",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "label": {
        "name": "yield",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;
